sycl/source/detail/graph_impl.hpp

```diff
@@ -1281,11 +1281,11 @@ class graph_impl {
   /// presence of the assume_buffer_outlives_graph property.
   bool MAllowBuffers = false;
 
-  /// List of nodes that must be added as extra dependencies to new nodes when
-  /// added to this graph.
-  /// This list is mainly used by barrier nodes which must be considered
-  /// as predecessors for all nodes subsequently added to the graph.
-  std::list<std::shared_ptr<node_impl>> MExtraDependencies;
+  /// Mapping from queues to barrier nodes. For each queue the last barrier
+  /// node recorded to the graph from the queue is stored.
+  std::map<std::weak_ptr<sycl::detail::queue_impl>, std::shared_ptr<node_impl>,
```

Comment:
Since we don't seem to care about the order, can this be turned into an unordered_map?

Reply:
I had this initially but it seems that `weak_ptr`s can't be used as a key in unordered containers https://stackoverflow.com/questions/13695640/how-to-make-a-c11-stdunordered-set-of-stdweak-ptr/30922307#30922307 The easiest way to solve this seemed to be an ordered map (I think because it avoids the need to compute a hash as it uses a comparison operation instead).

Reply:
I see, that's fair enough! I think using map is fine in this case.